### PR TITLE
Feat: PDP multiple page template

### PR DESCRIPTION
**Pull request by eduardoformiga**
Depends on:

- https://github.com/vtex/faststore/pull/2215
- https://github.com/vtex/vtex-cicd-platform/pull/572
- https://github.com/vtex/faststore/pull/2225

## What's the purpose of this pull request?

This PR aims to find the best PDP template from the CMS, based on the PDP slug and category tree.
[RFC Multiple Page template details](https://docs.google.com/document/d/1qQ573B9iVadSrqjvjSt8ExDu7-uxa3_COXlKvqkTQ8Q/edit)

vendemo account
Screenshot 2024-02-06 at 17 27 24

## How it works?

Prioritizing the following order:

1. A PDP template that matches the page slug (e.g. slug = /apple-magic-mouse/p).
2. A PDP template that matches the product subcategory (e.g. /department/category/subcategory).
3. A PDP template that matches the product category (e.g. /department/category).
4. A PDP template that matches the product department (e.g. /department).
5. If no matches are found, use the generic PDP template.

## How to test it?

1. localhost Initial setup:
You can change the tenant from CMS to the account `vendemo`. This store has all the PDPs templates created and published. Also, we changed the stored to `lojavivara`, for the sake of testing with a more complete catalog without harming the devs that are using `storeframework` nowadays.

Screenshot 2024-02-23 at 13 26 42

Screenshot 2024-02-23 at 13 26 02

2. Note that we have different templates created for PDPs with all scenarios mentioned above. So, use the following products to test

|Matching Type|URL|
|-|-|
|Slug - /anel-color-em-prata-925-com-topazio-sky-e-topazios-incolores-an00052637/p | http://localhost:3000/anel-color-em-prata-925-com-topazio-sky-e-topazios-incolores-an00052637/p |
|Subcategory - /vivara/joias/aneis/ |http://localhost:3000/anel-inside-em-prata-925-com-topazios-incolores-an00051429/p |
|Category - /vivara/joias/ |http://localhost:3000/pulseira-riviera-em-prata-925-com-espinelios-pu00016734/p |
|Department - /vivara/ |http://localhost:3000/relogio-tommy-hilfiger-masculino-couro-preto-1710542-to00004729/p |
|General|http://localhost:3000/corrente-para-coleira-life-for-pets-kl00000009/p |

### Starters Deploy Preview

This PR in starter uses the same configs mentioned. `vendemo` for CMS and `lojavivara` for API:
- https://github.com/vtex-sites/starter.store/pull/388

You can test using the same URLs:
|Matching Type|URL|
|-|-|
|Slug - /anel-color-em-prata-925-com-topazio-sky-e-topazios-incolores-an00052637/p | https://sfj-893f2c3--starter.preview.vtex.app/anel-color-em-prata-925-com-topazio-sky-e-topazios-incolores-an00052637/p |
|Subcategory - /vivara/joias/aneis/ |https://sfj-893f2c3--starter.preview.vtex.app/anel-inside-em-prata-925-com-topazios-incolores-an00051429/p |
|Category - /vivara/joias/ |https://sfj-893f2c3--starter.preview.vtex.app/pulseira-riviera-em-prata-925-com-espinelios-pu00016734/p |
|Department - /vivara/ |https://sfj-893f2c3--starter.preview.vtex.app/relogio-tommy-hilfiger-masculino-couro-preto-1710542-to00004729/p |
|General|https://sfj-893f2c3--starter.preview.vtex.app/corrente-para-coleira-life-for-pets-kl00000009/p |

## References

 Doc PR
- https://github.com/vtex/faststore/pull/2236
- [RFC Multiple Page template details](https://docs.google.com/document/d/1qQ573B9iVadSrqjvjSt8ExDu7-uxa3_COXlKvqkTQ8Q/edit)

### CODE REVIEW

File: packages/core/src/utils/utilities.ts [hunk 1]
```diff
@@ -11,3 +14,65 @@ export function textToKebabCase(text: string): string {
...
+ * @returns The best PDP template page for the slug
+ */
+export function findBestPDPTemplate(
+  pages: PDPContentType[],
+  slug: string,
+  product: ServerProductQueryQuery['product']
+) {
+  // productSlugAndCategoryTree with the prioritized order. [slug, subcategory tree, category tree, department]
+  const productSlugAndCategoryTree = product?.breadcrumbList?.itemListElement
+    ? [...product?.breadcrumbList?.itemListElement]
```
Comment (emersonlaurentino) [inline packages/core/src/utils/utilities.ts:60]: Why do you need to create a new array?

**Review by emersonlaurentino — APPROVED** [review]

File: packages/core/src/utils/utilities.ts (hunk 1 above)
Comment (eduardoformiga) [reply packages/core/src/utils/utilities.ts:60]: If we change the original object, the breadcrumb order will be affected.

File: packages/core/src/utils/utilities.ts [hunk 2]
```diff
@@ -11,3 +14,65 @@ export function textToKebabCase(text: string): string {
...
+ * 1. A PDP template that matches the product subcategory (e.g. /department/category/subcategory).
+ * 2. A PDP template that matches the product category (e.g. /department/category).
+ * 3. A PDP template that matches the product department (e.g. /department).
+ * 4. If no matches are found, use the generic PDP template.
+ *
+ * @param pages
+ * @param originalSlug
+ * @param product
+ * @returns The best PDP template page for the slug
+ */
```
Comment (lucasfp13) [inline packages/core/src/utils/utilities.ts:52]: Great to have these descriptions here 💯

File: packages/core/src/utils/utilities.ts [hunk 3]
```diff
@@ -11,3 +14,65 @@ export function textToKebabCase(text: string): string {
...
+
+/**
+ * Find the best PDP template from the CMS based on the slug or in the product category tree.
+ * Prioritizing the following order:
+ *
+ * 1. A PDP template that matches the page slug (e.g. slug = /apple-magic-mouse/p).
+ * 1. A PDP template that matches the product subcategory (e.g. /department/category/subcategory).
+ * 2. A PDP template that matches the product category (e.g. /department/category).
+ * 3. A PDP template that matches the product department (e.g. /department).
+ * 4. If no matches are found, use the generic PDP template.
```
Comment (hellofanny) [inline packages/core/src/utils/utilities.ts:46]:
```suggestion
 * 1. A PDP template that matches the page slug (e.g. slug = /apple-magic-mouse/p).
 * 2. A PDP template that matches the product subcategory (e.g. /department/category/subcategory).
 * 3. A PDP template that matches the product category (e.g. /department/category).
 * 4. A PDP template that matches the product department (e.g. /department).
 * 5. If no matches are found, use the generic PDP template.
```

File: packages/core/src/pages/[slug]/p.tsx [hunk 4]
```diff
@@ -242,6 +221,9 @@ export const getStaticProps: GetStaticProps<
     throw errors[0]
   }
 
+  const cmsPage: PDPContentType = await getPDP(slug, data.product, previewData)
+  console.log('🚀 ~ cmsPage:', cmsPage)
```
Comment (eduardoformiga) [inline packages/core/src/pages/[slug]/p.tsx:225]: 🎗️ remind
```suggestion
```

File: packages/core/src/server/cms/index.ts [hunk 5]
```diff
@@ -88,3 +33,36 @@ export type PageContentType = ContentData & {
...
+
+const isLocator = (x: any): x is Locator =>
+  typeof x.contentType === 'string' &&
+  (typeof x.releaseId === 'string' ||
+    typeof x.documentId === 'string' ||
+    typeof x.versionId === 'string')
+
+export const clientCMS = new ClientCMS({
+  workspace: config.api.workspace,
+  tenant: 'vendemo',
```
Comment (eduardoformiga) [inline packages/core/src/server/cms/index.ts:45]: 🎗️ 
```suggestion
  tenant: config.api.storeId,
```

File: packages/core/src/server/cms/pdp.ts [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+export const getPDP = async (
+  slug: string,
+  product: ServerProductQueryQuery['product'],
+  previewData: Locator
+) => {
+  console.log('🚀 ~ config.cms.data:', config.cms.data)
+  if (config.cms.data) {
+    const cmsData = JSON.parse(config.cms.data)
+    console.log('🚀 ~ cmsData:', cmsData)
```
Comment (eduardoformiga) [inline packages/core/src/server/cms/pdp.ts:31]: 🎗️ 
```suggestion
```

File: packages/core/src/server/cms/pdp.ts [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+} & PDPSettings
+
+export type PDPContentType = ContentData & PDPSettings
+
+export const getPDP = async (
+  slug: string,
+  product: ServerProductQueryQuery['product'],
+  previewData: Locator
+) => {
+  console.log('🚀 ~ config.cms.data:', config.cms.data)
```
Comment (eduardoformiga) [inline packages/core/src/server/cms/pdp.ts:28]: 🎗️ 
```suggestion
```

File: packages/core/src/server/cms/pdp.ts [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+  if (config.cms.data) {
+    const cmsData = JSON.parse(config.cms.data)
+    console.log('🚀 ~ cmsData:', cmsData)
+    const allPDPsFromCmsEnvData: PDPfromCmsEnvData[] = cmsData['pdp']
+    console.log('🚀 ~ allPDPsFromCmsEnvData:', allPDPsFromCmsEnvData)
```
Comment (eduardoformiga) [inline packages/core/src/server/cms/pdp.ts:33]: 🎗️ 
```suggestion
```

**Review by lariciamota — APPROVED** [review]
🎉

**Review by hellofanny — APPROVED** [review]
LGTM! 🥳

**Pull request merged**